src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,243 @@
+        private bool _disposed;
+
+        // Used by the Timer to call TryRelenish on ReplenishingRateLimiters
+        // We use a separate list to avoid running TryReplenish (which might be user code) inside our lock
+        // And we cache the list to amortize the allocation cost to as close to 0 as we can get
+        private List<Lazy<RateLimiter>>? _cachedLimiters = new();
+        private bool _cacheInvalid;
+        private TimerAwaitable _timer;
+        private Task _timerTask;
+
+        // Use the Dictionary as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _limiters;
+
+        public DefaultPartitionedRateLimiter(Func<TResource, RateLimitPartition<TKey>> partitioner,
+            IEqualityComparer<TKey>? equalityComparer = null)
+        {
+            _limiters = new Dictionary<TKey, Lazy<RateLimiter>>(equalityComparer);
+            _partitioner = partitioner;
+
+            // TODO: Figure out what interval we should use
+            _timer = new TimerAwaitable(TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(100));
+            _timerTask = RunTimer();
+        }
+
+        private async Task RunTimer()
+        {
+            _timer.Start();
+            while (await _timer)
+            {
+                Replenish(this);
+            }
+            _timer.Dispose();
+        }
+
+        public override int GetAvailablePermits(TResource resourceID)
+        {
+            return GetRateLimiter(resourceID).GetAvailablePermits();
+        }
+
+        protected override RateLimitLease AcquireCore(TResource resourceID, int permitCount)
+        {
+            return GetRateLimiter(resourceID).Acquire(permitCount);
+        }
+
+        protected override ValueTask<RateLimitLease> WaitAsyncCore(TResource resourceID, int permitCount, CancellationToken cancellationToken)
+        {
+            return GetRateLimiter(resourceID).WaitAsync(permitCount, cancellationToken);
+        }
+
+        private RateLimiter GetRateLimiter(TResource resourceID)
+        {
+            RateLimitPartition<TKey> partition = _partitioner(resourceID);
+            Lazy<RateLimiter>? limiter;
+            lock (Lock)
+            {
+                ThrowIfDisposed();
+                if (!_limiters.TryGetValue(partition.PartitionKey, out limiter))
+                {
+                    // Using Lazy avoids calling user code (partition.Factory) inside the lock
+                    limiter = new Lazy<RateLimiter>(() => partition.Factory(partition.PartitionKey));
+                    _limiters.Add(partition.PartitionKey, limiter);
+                    // Cache is invalid now
+                    _cacheInvalid = true;
+                }
+            }
+            return limiter.Value;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
```

Comment:
```suggestion
                _timerTask.GetAwaiter().GetResult();
                return;
```

Really we should just be setting a `_disposeTcs` or something so we don't complete after the timer stops but before all the `_limiters` are disposed.